imu_filter_madgwick/include/imu_filter_madgwick/imu_filter_ros.h

```diff
@@ -21,96 +21,83 @@
-    boost::shared_ptr<Synchronizer> sync_;
-
-    ros::Publisher rpy_filtered_debug_publisher_;
-    ros::Publisher rpy_raw_debug_publisher_;
-    ros::Publisher imu_publisher_;
-    tf2_ros::TransformBroadcaster tf_broadcaster_;
-
-    boost::shared_ptr<FilterConfigServer> config_server_;
-    ros::Timer check_topics_timer_;
-
-    // **** paramaters
-    WorldFrame::WorldFrame world_frame_;
-    bool use_mag_;
-    bool stateless_;
-    bool publish_tf_;
-    bool reverse_tf_;
-    std::string fixed_frame_;
-    std::string imu_frame_;
-    double constant_dt_;
-    bool publish_debug_topics_;
-    bool remove_gravity_vector_;
-    geometry_msgs::Vector3 mag_bias_;
-    double orientation_variance_;
-
-    // **** state variables
-    boost::mutex mutex_;
-    bool initialized_;
-    ros::Time last_time_;
-
-    // **** filter implementation
-    ImuFilter filter_;
-
-    // **** member functions
-    void imuMagCallback(const ImuMsg::ConstPtr& imu_msg_raw,
-                        const MagMsg::ConstPtr& mav_msg);
-
-    void imuCallback(const ImuMsg::ConstPtr& imu_msg_raw);
-
-    void publishFilteredMsg(const ImuMsg::ConstPtr& imu_msg_raw);
-    void publishTransform(const ImuMsg::ConstPtr& imu_msg_raw);
-
-    void publishRawMsg(const ros::Time& t,
-                       float roll, float pitch, float yaw);
-
-    void reconfigCallback(FilterConfig& config, uint32_t level);
-    void checkTopicsTimerCallback(const ros::TimerEvent&);
-};
-
-#endif // IMU_FILTER_IMU_MADWICK_FILTER_ROS_H
+public:
+  explicit ImuFilterMadgwickRos(const rclcpp::NodeOptions &options);
+
+private:
+  std::shared_ptr<ImuSubscriber> imu_subscriber_;
+  std::shared_ptr<MagSubscriber> mag_subscriber_;
+  std::shared_ptr<Synchronizer> sync_;
+
+  rclcpp::Publisher<MagMsg>::SharedPtr mag_republisher_;
+  rclcpp::Publisher<MagVectorMsg>::SharedPtr rpy_filtered_debug_publisher_;
+  rclcpp::Publisher<MagVectorMsg>::SharedPtr rpy_raw_debug_publisher_;
+  rclcpp::Publisher<ImuMsg>::SharedPtr imu_publisher_;
+  tf2_ros::TransformBroadcaster tf_broadcaster_;
+
+  rclcpp::TimerBase::SharedPtr check_topics_timer_;
+
+  // Subscription for parameter change
+  rclcpp::AsyncParametersClient::SharedPtr parameters_client_;
+  rclcpp::Subscription<rcl_interfaces::msg::ParameterEvent>::SharedPtr parameter_event_sub_;
+
+  // **** paramaters
+  WorldFrame::WorldFrame world_frame_;
+  bool use_mag_;
+  bool stateless_;
+  bool publish_tf_;
+  bool reverse_tf_;
+  std::string fixed_frame_;
+  std::string imu_frame_;
+  double constant_dt_;
+  bool publish_debug_topics_;
+  bool remove_gravity_vector_;
+  geometry_msgs::msg::Vector3 mag_bias_;
+  double orientation_variance_;
+
+  // **** state variables
+  std::mutex mutex_;
+  bool initialized_;
+  rclcpp::Time last_time_;
+
+  // **** filter implementation
+  ImuFilter filter_;
+
+  // **** member functions
+  void imuMagCallback(const ImuMsg::SharedPtr imu_msg_raw, const MagMsg::SharedPtr mav_msg);
```

Comment:
Could you use `ConstSharedPtr` in all the callbacks instead? This ensures that we use the messages read-only, and either prevents a useless copy or potentially incorrect behavior. I think this should work (haven't tested it yet):

* instead of: `const ImuMsg::SharedPtr imu_msg_raw` ...
* use this: `ImuMsg::ConstSharedPtr imu_msg_raw` (notice that the `const` has to be removed)

Reply:
Done

Reply:
You forgot the `imuCallback` :)